
<center>
    <font size="+2">
    <b>Artificial Intelligence<br></b>
        <b>Project 1: 8-Puzzle</b>
    </font>
</center>


In this project you will implement the informed version of A* to solve the 8 puzzle. You will implement 2 different heuristic measures as defined in the slides and report the results. You will also develop your own heuristic measure to solve this problem. For this you will need to implement these heuristic measures in the class State. You will also have to research how to create a priority queue in python and how you can make the class `State` comparable. 

1.1 Add a variable g that will hold the cumulative backward cost

1.2 Implement the Tiles heuristic measure to compute h

1.3 Implement the Manhattan heuristic to compute h

1.4 Test both implementations and report the values for nodes visited and explored. 

1.5 Design you own heuristic measure and implement it (It has to be consistent)

In [1]:
import numpy as np
class State:
    def __init__(self,s, zero, prev=None, g=0):
        self.s=s
        self.prev=prev
        self.zero=zero
        self.hash=None
        
        self.g=g
        
        self.tiles_heuristic=self.h_heuristic()
        self.manhattan_heuristic=self.m_heuristic()

        
        self.f_tiles=self.g+self.tiles_heuristic
        self.f_manhattan=self.g+self.manhattan_heuristic
        self.f_kazem=self.g+self.manhattan_heuristic+self.tiles_heuristic
        
        
    def __hash__(self):
        if self.hash:
            return self.hash
        s=0
        for i in range(8,-1,-1):
            s=10*s+self.s[i]
        self.hash=s
        return s
        
    #1,5,6
    #2,0,3
    #4,7,8
    
    #[1,5,6,2,0,3,4,7,8] zero=4
    def Successors(self):
        succ=[]
        s=self.s
        zero=self.zero
        if (zero-1)>=0 and ((zero-1)//3==zero//3):
            tmp=s.copy()
            tmp[zero-1],tmp[zero]=s[zero],s[zero-1]
            succ.append(State(tmp,zero-1,self,self.g+1))
        if (zero+1<9) and ((zero+1)//3==zero//3):
            tmp=s.copy()
            tmp[zero+1],tmp[zero]=s[zero],s[zero+1]
            succ.append(State(tmp,zero+1,self,self.g+1))
        if (zero-3)>=0:
            tmp=s.copy()
            tmp[zero-3],tmp[zero]=s[zero],s[zero-3]
            succ.append(State(tmp,zero-3,self,self.g+1))
        if (zero+3)<=8:
            tmp=s.copy()
            tmp[zero+3],tmp[zero]=s[zero],s[zero+3]
            succ.append(State(tmp,zero+3,self,self.g+1))
        
        return succ
    
    def IsGoal(self):
        solution=[1,2,3,4,5,6,7,8,0]
        s=self.s
        for i in range(9):
            if s[i]!=solution[i]:
                return False
        return True
    
    def __eq__(self, other):
        for i in range(9):
            if self.s[i]!=other.s[i]:
                return False
        return True
    
    def h_heuristic(self):
        goal=[1,2,3,4,5,6,7,8,0]
        h=0
        for i in range(9):
            if self.s[i]!=goal[i]:
                h=h+1
        return h
    
    def m_heuristic(self):
        goal=[1,2,3,4,5,6,7,8,0]
        m=0
        for i in range(len(self.s)):
            stateIndex= self.s.index(self.s[i])
            stateX=stateIndex//3
            stateY=stateIndex%3

            goalIndex = goal.index(self.s[i])
            goalX=goalIndex//3
            goalY=goalIndex%3

            xDif= abs(stateX-goalX)
            yDif= abs(stateY-goalY)

            m=m+(xDif+yDif)
        
        return m
    
    
    def __str__(self):
        
        tmp=np.array(self.s)
        return str(tmp.reshape(3,3))

In [2]:
def Solution(state):
    path=[]
    length=0
    while state:
        path.append(str(state))
        state=state.prev
        length=length+1
    return ('\n'.join(path), length)

# A*

Change the following code to implement the A* algorithm. 

In [9]:
start=State([7,2,4,5,0,6,8,3,1],4)

visited=set()
solved=False
from collections import deque
#q = deque() 
q=[]
q.append([[start],[0]])
i=0



def sorKey(A):
    return A[1]

"""

# A* Algorithm with Tiles Heuristic
print("A* Algorithm with Tiles Heuristic")
while len(q) > 0:
    i=i+1
    current=q.pop()
    current = current[0]
    current = current[0]
    if current.IsGoal():
        path, length= Solution(current)
        print(path)
        print("Moves: "+str(length))
        solved=True
        break
    if current in visited:
        continue
    visited.add(current)
    for tmp in current.Successors():
        #q.append(tmp)
        q.append([[tmp],[tmp.f_tiles]])
        q.sort(reverse=True,key = sorKey) 
        

        
# A* Algorithm with Manhatten Heuristic
print("A* Algorithm with Manhatten Heuristic")
while len(q) > 0:
    i=i+1
    current=q.pop()
    current = current[0]
    current = current[0]
    if current.IsGoal():
        path, length= Solution(current)
        print(path)
        print("Moves: "+str(length))
        solved=True
        break
    if current in visited:
        continue
    visited.add(current)
    for tmp in current.Successors():
        #q.append(tmp)
        q.append([[tmp],[tmp.f_manhattan]])
        q.sort(reverse=True,key = sorKey) 

        
"""
        
# A* Algorithm with Kazem Heuristic
print("A* Algorithm with Kazem Heuristic")
while len(q) > 0:
    i=i+1
    current=q.pop()
    current = current[0]
    current = current[0]
    if current.IsGoal():
        path, length= Solution(current)
        print(path)
        print("Moves: "+str(length))
        solved=True
        break
    if current in visited:
        continue
    visited.add(current)
    for tmp in current.Successors():
        #q.append(tmp)
        q.append([[tmp],[tmp.f_kazem]])
        q.sort(reverse=True,key = sorKey)  
        

        
if not solved: 
    print("No Solution")
    

print("Total Backward Cost: "+str(current.g))
print("Visited states: "+str(len(visited)))
print("Explored: "+str(i))



A* Algorithm with Kazem Heuristic
[[1 2 3]
 [4 5 6]
 [7 8 0]]
[[1 2 3]
 [4 5 0]
 [7 8 6]]
[[1 2 3]
 [4 0 5]
 [7 8 6]]
[[1 2 3]
 [0 4 5]
 [7 8 6]]
[[0 2 3]
 [1 4 5]
 [7 8 6]]
[[2 0 3]
 [1 4 5]
 [7 8 6]]
[[2 4 3]
 [1 0 5]
 [7 8 6]]
[[2 4 3]
 [0 1 5]
 [7 8 6]]
[[2 4 3]
 [7 1 5]
 [0 8 6]]
[[2 4 3]
 [7 1 5]
 [8 0 6]]
[[2 4 3]
 [7 0 5]
 [8 1 6]]
[[2 4 3]
 [7 5 0]
 [8 1 6]]
[[2 4 0]
 [7 5 3]
 [8 1 6]]
[[2 0 4]
 [7 5 3]
 [8 1 6]]
[[0 2 4]
 [7 5 3]
 [8 1 6]]
[[7 2 4]
 [0 5 3]
 [8 1 6]]
[[7 2 4]
 [5 0 3]
 [8 1 6]]
[[7 2 4]
 [5 3 0]
 [8 1 6]]
[[7 2 4]
 [5 3 6]
 [8 1 0]]
[[7 2 4]
 [5 3 6]
 [8 0 1]]
[[7 2 4]
 [5 0 6]
 [8 3 1]]
Moves: 21
Total Backward Cost: 20
Visited states: 126
Explored: 204


## Fill the values here

Nodes Visited (Tiles): 2787

Nodes Visited (Manhattan): 436

Nodes Visited (Kazem): 126

Nodes Explored (Tiles): 4442

Nodes Explored (Manhattan): 703

Nodes Explored (Kazem): 204

Name 1 : Kazem Al-Seblani

Name 2 :

Name 3 :

## How "Kazem" heuristic works:
while tiles heuristic were better than BFS, and Manhattan were even better than tiles, I tried to get the advantage of both together.
my heuristic took into consideration the difference between tiles as well as the distance between the current item and where it should be.

that's why it gave results better than Manhattan
